core/src/main/java/ai/timefold/solver/core/impl/domain/valuerange/descriptor/CompositeValueRangeDescriptor.java

```diff
@@ -9,88 +9,69 @@
 import ai.timefold.solver.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
 
+import org.jspecify.annotations.NullMarked;
+
 /**
  * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
  */
-public class CompositeValueRangeDescriptor<Solution_> extends AbstractValueRangeDescriptor<Solution_>
-        implements EntityIndependentValueRangeDescriptor<Solution_> {
+@NullMarked
+public final class CompositeValueRangeDescriptor<Solution_> extends AbstractValueRangeDescriptor<Solution_> {
 
-    protected final List<ValueRangeDescriptor<Solution_>> childValueRangeDescriptorList;
-    protected boolean entityIndependent;
+    private final boolean canExtractValueRangeFromSolution;
+    private final boolean acceptsNullInValueRange;
+    private final List<ValueRangeDescriptor<Solution_>> childValueRangeDescriptorList;
 
-    public CompositeValueRangeDescriptor(
-            GenuineVariableDescriptor<Solution_> variableDescriptor, boolean addNullInValueRange,
+    public CompositeValueRangeDescriptor(GenuineVariableDescriptor<Solution_> variableDescriptor,
             List<ValueRangeDescriptor<Solution_>> childValueRangeDescriptorList) {
-        super(variableDescriptor, addNullInValueRange);
+        super(variableDescriptor);
         this.childValueRangeDescriptorList = childValueRangeDescriptorList;
-        entityIndependent = true;
+        var canExtractFromSolution = true;
+        var acceptsNull = false;
         for (var valueRangeDescriptor : childValueRangeDescriptorList) {
             if (!valueRangeDescriptor.isCountable()) {
                 throw new IllegalStateException(
                         "The valueRangeDescriptor (%s) has a childValueRangeDescriptor (%s) with countable (%s)."
                                 .formatted(this, valueRangeDescriptor, valueRangeDescriptor.isCountable()));
             }
-            if (!valueRangeDescriptor.isEntityIndependent()) {
-                entityIndependent = false;
-            }
+            canExtractFromSolution = canExtractFromSolution && valueRangeDescriptor.canExtractValueRangeFromSolution();
+            acceptsNull = acceptsNull || valueRangeDescriptor.acceptsNullInValueRange();
         }
+        this.canExtractValueRangeFromSolution = canExtractFromSolution;
+        this.acceptsNullInValueRange = acceptsNull;
     }
 
     @Override
-    public boolean isCountable() {
-        return true;
+    public boolean canExtractValueRangeFromSolution() {
+        return canExtractValueRangeFromSolution;
     }
 
     @Override
-    public boolean isEntityIndependent() {
-        return entityIndependent;
+    public boolean isCountable() {
+        return true;
     }
 
     @Override
-    public <Value_> ValueRange<Value_> extractValueRange(Solution_ solution, Object entity) {
-        return innerExtractValueRange(solution, entity);
+    public boolean acceptsNullInValueRange() {
+        return acceptsNullInValueRange;
```

Comment:
I was a bit surprised that the composite range does not deal with null.
(The code was effectively always putting false here.)

What I don't really understand is ... if the value range is composite... how does it ever get a null in there? IMO this code should go over all the child ranges and remember if either one of them accepts null.

I made that happen, but (new) tests that expect the previous behavior now fail. I am not entirely sure that the tests are correct.